pkg/controller/plan/adapter/ocp/builder.go

```diff
@@ -0,0 +1,469 @@
+		var size resource.Quantity
+		var volName, volNamespace string
+		if vol.PersistentVolumeClaim != nil {
+			pvc := &core.PersistentVolumeClaim{}
+			err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: vmRef.Namespace, Name: vol.PersistentVolumeClaim.ClaimName}, pvc)
+			if err != nil {
+				return nil, liberr.Wrap(err)
+			}
+			volName = pvc.Name
+			volNamespace = pvc.Namespace
+			size = pvc.Spec.Resources.Requests["storage"]
+		} else if vol.DataVolume != nil {
+			pvc := &core.PersistentVolumeClaim{}
+			err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: vmRef.Namespace, Name: vol.DataVolume.Name}, pvc)
+			if err != nil {
+				return nil, liberr.Wrap(err)
+			}
+			volName = pvc.Name
+			volNamespace = pvc.Namespace
+			size = pvc.Spec.Resources.Requests["storage"]
+		}
+
+		mB := size.Value() / 1024 / 1024
+		list = append(
+			list,
+			&planapi.Task{
+				Name: fmt.Sprintf("%s/%s", volNamespace, volName),
+				Progress: libitr.Progress{
+					Total: mB,
+				},
+				Annotations: map[string]string{
+					"unit": "MB",
+				},
+			})
+	}
+
+	return
+}
+
+// TemplateLabels implements base.Builder
+func (r *Builder) TemplateLabels(vmRef ref.Ref) (labels map[string]string, err error) {
+	return nil, nil
+}
+
+// VirtualMachine implements base.Builder
+func (r *Builder) VirtualMachine(vmRef ref.Ref, object *cnv.VirtualMachineSpec, persistentVolumeClaims []core.PersistentVolumeClaim) error {
+	// Get vm export
+	vmExport := &export.VirtualMachineExport{}
+	err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: vmRef.Namespace, Name: vmRef.Name}, vmExport)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	sourceVm, err := r.generateVmFromDefinition(vmExport)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	targetVMspec := sourceVm.Spec
+
+	// TODO: move logic to mapDisks
+	pvcMap := make(map[string]*core.PersistentVolumeClaim)
+	for i := range persistentVolumeClaims {
+		pvc := &persistentVolumeClaims[i]
+		if source, ok := pvc.Annotations[planbase.AnnDiskSource]; ok {
+			pvcMap[source] = pvc
+		}
+	}
+
+	diskMap := make(map[string]*cnv.Disk)
+	for i := range sourceVm.Spec.Template.Spec.Domain.Devices.Disks {
+		disk := &sourceVm.Spec.Template.Spec.Domain.Devices.Disks[i]
+		// Find matching volume
+		// TODO: make this better
+		for _, vol := range sourceVm.Spec.Template.Spec.Volumes {
+			if vol.PersistentVolumeClaim != nil {
+				if vol.Name == disk.Name {
+					key := pvcSourceName(vmRef.Namespace, vol.PersistentVolumeClaim.ClaimName)
+					if _, ok := pvcMap[key]; ok {
+						diskMap[key] = disk
+						continue
+					}
+				}
+			} else if vol.DataVolume != nil {
+				if vol.Name == disk.Name {
+					key := pvcSourceName(vmRef.Namespace, vol.DataVolume.Name)
+					if _, ok := pvcMap[key]; ok {
+						diskMap[key] = disk
+						continue
+					}
+				}
+			}
+		}
+	}
+
+	// Clear original disks and volumes, will be required for other mapped devices later
```

Comment:
so maybe we should change generateVmFromDefinition to return a target vm with these empty collections?